File: boj/4_graph/10451.cpp
#include <cstdio>

int T, N;
int next[1001];
bool visit[1001];

void dfs(int v) {
    while (!visit[v]) {
        visit[v] = true;
        v = next[v];
    }
}

int main() {
    scanf("%d", &T);
    while (T--) {
        scanf("%d", &N);
        for (int v = 1; v <= N; v++) {
            scanf("%d", &next[v]);
            visit[v] = false;
        }

        int ans = 0;
        for (int v = 1; v <= N; v++) {
            if (!visit[v]) {
                ++ans, dfs(v);
            }
        }
        printf("%d\n", ans);
    }
}

File: 2023/4-graph-230527_0529/10451.py
import sys
input = lambda: sys.stdin.readline().rstrip()
print = lambda *args: sys.stdout.write(f"{' '.join(map(str, args))}\n")
####

def dfs(adj, visit, v):
	visit[v] = True
	for w in adj[v]:
		if visit[w] == False:
			dfs(adj, visit, w)

T = int(input())
for _ in range(T):
	N = int(input())
	A = [0] + list(map(int, input().split()))

	adj = [ [] for _ in range(N + 1) ]
	for v in range(1, N + 1):
		adj[v].append(A[v]); adj[A[v]].append(v)
	
	visit = [ False for _ in range(N + 1) ]
	
	ans = 0
	for v in range(1, N + 1):
		if visit[v] == False:
			ans += 1
			dfs(adj, visit, v)
	print(ans)